src/interceptors/WebSocket/WebSocketController.ts

```diff
@@ -0,0 +1,347 @@
+function createWebSocketProxy(
+  ws: WebSocket,
+  options: {
+    onSend: (data: WebSocketData, next: NextFunction<void>) => void
+    onReceived: (data: WebSocketData, next: NextFunction<void>) => void
+  }
+): WebSocket {
+  const proxy = createProxy(ws, {
+    methodCall([methodName, args], next) {
+      switch (methodName) {
+        case 'send': {
+          options.onSend(args[0], next)
+          break
+        }
+
+        case 'dispatchEvent': {
+          const [event] = args
+
+          if (event instanceof MessageEvent && event.type === 'message') {
+            options.onReceived(event.data, next)
+            return
+          }
+
+          return next()
+        }
+
+        default: {
+          return next()
+        }
+      }
+    },
+  })
+
+  return proxy
+}
+
+class WebSocketOverride extends EventTarget implements WebSocket {
+  static readonly CONNECTING = WebSocket.CONNECTING
+  static readonly OPEN = WebSocket.OPEN
+  static readonly CLOSING = WebSocket.CLOSING
+  static readonly CLOSED = WebSocket.CLOSED
+  readonly CONNECTING = WebSocket.CONNECTING
+  readonly OPEN = WebSocket.OPEN
+  readonly CLOSING = WebSocket.CLOSING
+  readonly CLOSED = WebSocket.CLOSED
+
+  public url: string
+  public protocol: string
+  public extensions: string
+  public bufferedAmount: number
+  public binaryType: BinaryType
+  public readyState: number
+
+  private _url: string | URL
+  private _protocols?: string | Array<string>
+  private _events: Map<string, Set<EventListenerOrEventListenerObject>>
+  private _onopen: ((this: WebSocket, event: Event) => void) | null = null
+  private _onmessage: ((this: WebSocket, event: MessageEvent) => void) | null =
+    null
+  private _onclose: ((this: WebSocket, event: CloseEvent) => void) | null = null
+  private _onerror: ((this: WebSocket, event: Event) => void) | null = null
+
+  constructor(url: URL | string, protocols?: string | Array<string>) {
+    super()
+
+    this._url = url
+    this._protocols = protocols
+    this._events = new Map()
+
+    this.url = url.toString()
+    this.protocol = protocols ? protocols[0] : 'ws'
+    this.extensions = ''
+    this.bufferedAmount = 0
+    this.binaryType = 'arraybuffer'
+
+    this.readyState = this.CONNECTING
+  }
+
+  public unwrap(WebSocketClass: typeof WebSocket): WebSocket {
+    const ws = new WebSocketClass(this._url, this._protocols)
+
+    // Forward any event listeners added to the override
+    // to the original WebSocket instance.
+    this._events.forEach((listeners, event) => {
+      listeners.forEach((listener) => {
+        ws.addEventListener(event, listener)
+      })
+    })
+
+    return ws
+  }
+
+  public send(data: string | ArrayBufferLike | Blob | ArrayBufferView): void {
+    if (this.readyState === this.CONNECTING) {
+      this.close()
+      throw new Error('InvalidStateError')
+    }
+
+    if (this.readyState === this.CLOSING || this.readyState === this.CLOSED) {
+      this.bufferedAmount += getDataSize(data)
```

Comment:
This is incorrect. Nothing must happen if sending during closing/closed. Data is buffered on regular `send()` because the actual data transfer happens on the next tick.